File: tree/112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        
        return self.dfs(root, sum)
    
    def dfs(self, root, left):
        if not root:
            return False
        if not root.left and not root.right:
            if root.val == left:
                return True
            else:
                return False
        else:
            return self.dfs(root.left, left - root.val) or self.dfs(root.right, left - root.val)
    

File: tree/112.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
    if root == nil {
        return false
    }
    return dfs(root, sum)
}

func dfs(root *TreeNode, left int) bool{
    if root.Left == nil && root.Right == nil {
        return left == root.Val
    }
    if root.Left != nil && root.Right != nil {
        return dfs(root.Left, left - root.Val) || dfs(root.Right, left - root.Val)
    }
    if root.Left != nil {
        return dfs(root.Left, left - root.Val)
    }
    if root.Right != nil {
        return dfs(root.Right, left - root.Val)
    }
    return false
}
